java/arcs/core/testutil/Fuzzing.kt

```diff
@@ -0,0 +1,344 @@
+ * setup:
+ *
+ * fun invariant_balanceCantGoNegative(initialBalance: A<Int>, transactions: A<List<Transaction>>)
+ * {
+ *   // SETUP
+ *   val theInitialBalance = initialBalance()
+ *
+ *   // PRECONDITION
+ *   assertThat(theInitialBalance).isLargerThan(0)
+ *
+ *   // BODY
+ *   val account = Account(theInitialBalance)
+ *   val transactionsApplied = transactions().map { account.apply(it) }
+ *
+ *   // POSTCONDITION
+ *   assertThat(account.balance).isLargerThan(0)
+ * }
+ *
+ * There are a few reasons to pass the randomness in as a parameter directly, rather than (a)
+ * generating random values inside the invariant or (b) generating random values in the invariant
+ * caller:
+ *   A) it promotes reuse & componentization of generators
+ *   B) it allows invariant functions to be used both for unit testing and fuzz testing
+ *   C) it lets constraints be built into generators rather than baked into each test
+ *   D) it lets dependencies between input values be expressed (to a degree) - see the 'T' class
+ *      below for more details.
+ *
+ * (3) reuse or build generators for the non-required structure
+ *
+ * Here's a simple generator for non-negative initial balances:
+ *
+ * class NonNegativeIntGenerator(s: Seed): A<Int> {
+ *   override operator fun invoke(): Int = s.nextInRange(0, Int.MAX_VALUE)
+ * }
+ *
+ * 's' is our source of randomness. In this case, we can use it directly to get an
+ * integer value.
+ *
+ * Let's assume we have a simple generator for Transactions too. Note that this generator could
+ * build on NonNegativeIntGenerator: both withdrawals and deposits would be expected to be
+ * non-negative. The pattern for this is to pass generators in as constructor parameters
+ * - see ListOf below, or arcs.core.data.PlanParticleGenerator for examples.
+ *
+ * (4) set up a fuzz test using runFuzzTest, your generators, and your invariant. runFuzzTest will
+ * establish a random seed for you, and print the seed value if your tests throws.
+ *
+ * @Test
+ * fuzz_balanceCantGoNegative = runFuzzTest {
+ *   invariant_balanceCantGoNegative(NonNegativeIntGenerator(it), TransactionGenerator(it))
+ * }
+ *
+ * Note that you can also use the invariant for unit tests (the Value class provides an
+ * implementation of 'A' for a single explicit value):
+ *
+ * @Test
+ * aLargeTransactionWontApply() {
+ *   invariant_balanceCantGoNegative(Value(100), Value(listOf(Withdrawal(200))))
+ * }
+ *
+ * It's also easy to use invariant functions to test a range of input conditions:
+ *
+ * @Test
+ * aRangeOfStartingBalances_MaintainInvariant() {
+ *   ...
+ *   listOf(4, 100, 1000, 20, 40, 50).forEach {
+ *     invariant_balanceCantGoNegative(Value(it), transactions)
+ *   }
+ * }
+ *
+ * Finally, you can capture failing fuzz runs explicitly as regression tests:
+ *
+ * @Test
+ * regression_balanceCantGoNegative() {
+ *   var seed = SeededRandom(-35623452)
```

Comment:
Eventually, that would be best. It's totally out of scope for this change though, as it requires threading a bunch of infrastructure through the generators.

For now, this is good enough - the only way that the generator function would change is if Kotlin (and therefore Java) changes how it generates randomness.

Reply:
it would also change if we change the generator. +1 to not doing in this change, but I would add a TODO and make sure we address it soon.

Reply:
I might be missing something, but it looks like there's basically two layers to each test:

1. The outer layer is responsible for generating different versions of the input data, and passing it through to the inner layer.
2. The inner layer is responsible for running a test on some given input data.

If we know the seed that generated the problematic input data, then what is preventing us from capturing that input data itself instead of capturing the seed? Then you replay the inner layer of the test using that captured input data.

I'm imagining something like this:

```
@Test
fun fuzzTest_balanceCantGoNegative() {
  // "Outer layer" (generates the input data).
  runFuzzTest(NonNegativeIntGenerator, TransactionGenerator) { initialBalance, transactions ->
    // "Inner layer" (runs the test with given data).
    test_balanceCantGoNegative(initialBalance, transactions)
  }
}

/** Regression test for b/123456... */
@Test
fun regression_balanceCantGoNegative() {
  test_balanceCantGoNegative(initialBalance = 123, transactions = ...)
}

fun test_balanceCantGoNegative(initialBalance: Int, transactions: List<Transaction>) {
   ...
}
```

Reply:
I would expect for meaningful regression tests that we'd do precisely that, Cameron.

However, to do so you need to be able to execute the test with the right random seed, so you can inspect and extract the inputs. Also, some inputs are potentially going to be very large or difficult to construct.